Introduce srml/im-online

cmichi opened the pull request:
Addresses https://github.com/paritytech/substrate/issues/2719.

There are not yet any tests, I'll add those in a follow-up PR. The PR is mostly here to quickly get this into the review pipeline and master.

I currently keep only the `Vec<u8>` version of `Multiaddr` and `PeerId.to_string()` in the runtime, since `parity-multiaddr` doesn't compile for wasm currently and `libp2p::core::PeerId` doesn't support encoding/serializing yet.

cmichi commented on srml/im-online/src/lib.rs:
```diff
@@ -0,0 +1,305 @@
+				match gossip::<T>(block_number) {
+					Ok(_) => {},
+					Err(_) => {
+						print("Error in offchain worker!");
```
I'm not sure about the proper handling here. Maybe it makes sense to only print something out for more serious things?

These are the possible errors:
```
enum OffchainErr {
	DecodeAuthorityId,
	ExtrinsicCreation,
	FailedSigning,
	LocalNetworkState,
	SubmitTransaction,
}
```

tomusdrw replied:
Would be best to just debug-print the entire error. I think we will need some logging from within offchain workers anyhow. Can you implement `sr_io::Printable` for the error and just do `print(err)` here?

tomaka commented on core/service/Cargo.toml:
```diff
@@ -32,6 +32,7 @@ transaction_pool = { package = "substrate-transaction-pool", path = "../../core/
 rpc = { package = "substrate-rpc-servers", path = "../../core/rpc-servers" }
 tel = { package = "substrate-telemetry", path = "../../core/telemetry" }
 offchain = { package = "substrate-offchain", path = "../../core/offchain" }
+parity-multiaddr = { package = "parity-multiaddr", version = "0.5.0" }
```
I don't think you're using this dependency, and if you are, `multiaddr` is reexported by `network`.

tomaka commented on core/offchain/src/api.rs:
```diff
@@ -158,6 +170,32 @@ impl<Storage, KeyProvider> OffchainExt for Api<Storage, KeyProvider> where
+			external_addresses: network_state.external_addresses(),
+
+			// TODO derive `Serialize` for `libp2p::core::PeerId`
+			peer_id: network_state.peer_id().to_string(),
```
```suggestion
			peer_id: network_state.peer_id().to_base58(),
```

`PeerId` is in this weird spot where the serialization option is not obvious, so you have to explicitly call `to_base58()` if you want base58, or `as_bytes()` if you want bytes.

tomaka commented on core/offchain/Cargo.toml:
```diff
@@ -16,6 +16,10 @@ parking_lot = "0.8.0"
+network = { package = "substrate-network", path = "../../core/network" }
+serde = { version = "1.0", features = ["derive"] }
+bincode = "1"
+parity_multiaddr = { package = "parity-multiaddr", version = "0.5.0" }
```
`multiaddr` can be accessed through `network::multiaddr`.

```suggestion
```

tomaka commented on core/offchain/src/api.rs:
```diff
@@ -33,6 +33,10 @@ use runtime_primitives::{
 };
 use transaction_pool::txpool::{Pool, ChainApi};
+use network::NetworkStateInfo;
+use parity_multiaddr::Multiaddr;
```
```suggestion
use network::multiaddr::Multiaddr;
```

tomaka commented:
Also, `parity-multiaddr` should normally compile fine for WASM. It's used in my WASM browser demo.

tomaka commented on core/sr-primitives/src/testing.rs:
```diff
@@ -44,6 +44,12 @@ impl TypedKey for UintAuthorityId {
 
+impl AsRef<[u8]> for UintAuthorityId {
+	fn as_ref(&self) -> &[u8] {
+		unsafe { &std::mem::transmute::<_, &[u8; 4]>(&self.0)[..] }
```
Let's be **extremely** careful when using transmute. This code definitely isn't correct.

bkchr replied:
Besides `u64` being `[u8;8]`, I don't see any problem here?

bkchr replied:
And casting to `[u8;4]` is valid.

bkchr replied:
Yeah, my fault. We should use `slice::from_raw_parts` here to get a correct fat pointer.

cmichi commented:
@tomaka 

> Also, parity-multiaddr should normally compile fine for WASM. It's used in my WASM browser demo.

Ok, thanks. I'll keep the encoded version in wasm for now, since we will probably never need the info in WASM, but only "outside". So we can save the dependency until there is a need to have the encoded object in WASM.

I addressed your comments, can you please take a look again?

tomusdrw commented on srml/im-online/src/lib.rs:
```diff
@@ -0,0 +1,305 @@
+				if signature.len() != 64 {
+					return TransactionValidity::Invalid(ApplyError::BadSignature as i8);
+				}
+				let mut array = [0; 64];
```
Maybe something like this to have a proper name instead of `array`?
```rust
let signature = {
  let mut array = [0; 64]
  array.copy_from_slice(&signature); // we already know the len is ok, since we have a check earlier
  array
};
```

tomusdrw commented on srml/im-online/src/lib.rs:
```diff
@@ -0,0 +1,305 @@
+					return TransactionValidity::Invalid(ApplyError::BadSignature as i8);
+				}
+
+				let transaction_tag = heartbeat.encode();
```
You `encode` it for the second time here - please make sure we do it only once.

tomusdrw commented on srml/im-online/src/lib.rs:
```diff
@@ -0,0 +1,305 @@
+
+	fn validate_unsigned(call: &Self::Call) -> srml_support::unsigned::TransactionValidity {
+		match call {
+			Call::heartbeat(heartbeat, signature) => {
```
We need additional check to make sure the `hearbeat` was not already sent by that `authority_id`. Most likely using `is_online_in_current_era`

tomusdrw commented on srml/im-online/src/lib.rs:
```diff
@@ -0,0 +1,305 @@
+	type Call = Call<T>;
+
+	fn validate_unsigned(call: &Self::Call) -> srml_support::unsigned::TransactionValidity {
+		match call {
```
maybe just do `if let Call::hearbeat(hearbeat, signature) = call {` since you only want one value. This will also lower the indentation level.

tomusdrw commented on srml/im-online/src/lib.rs:
```diff
@@ -0,0 +1,305 @@
+
+			let gossip_at = <GossipAt<T>>::get();
+			let now = <system::Module<T>>::block_number();
+			if gossip_at == now {
```
We don't really have guarantees that offchain worker will run for that particular block (You might actually be offline).

I think it's better to check that `gossip_at < now` and have additional Offchain DB entry that says that we already submitted the statement for current session.

tomusdrw commented on srml/im-online/src/lib.rs:
```diff
@@ -0,0 +1,305 @@
+			}
+
+			let gossip_at = <GossipAt<T>>::get();
+			let now = <system::Module<T>>::block_number();
```
We already have `block_number`, I don't think we need to query `system`. Alternative we can get rid of this parameter for offchain workers.

tomusdrw commented on srml/im-online/src/lib.rs:
```diff
@@ -0,0 +1,305 @@
+			let gossip_at = <GossipAt<T>>::get();
+			let now = <system::Module<T>>::block_number();
+			if gossip_at == now {
+				match gossip::<T>(block_number) {
```
should be `gossip_at` after we change the conditions.

mxinden commented on srml/im-online/src/lib.rs:
```diff
@@ -0,0 +1,305 @@
+				}
+			}
+
+			let gossip_at = <GossipAt<T>>::get();
```
If I understand correctly `gossip_at` is set only once in the initial chain specification. Thus the code below would only ever run once. Shouldn't this run multiple times each session? Am I missing something?

cmichi replied:
If you look at the implementation of `fn new_session()` (the hook which will be called by `srml-session` once the session changes):
```
let now = <system::Module<T>>::block_number();
<GossipAt<T>>::put(now);
```
So the value is updated each time the session changes, resulting in the code running for each session change.

mxinden replied:
Thanks @cmichi.

pepyakin commented on core/sr-io/without_std.rs:
```diff
@@ -385,6 +385,20 @@ pub mod ext {
+		///
+		/// # Returns
+		///
+		/// - The encoded `Result<offchain::LocalNetworkState, ()>`.
```
That's a good question and I cannot answer it properly because I wasn't the most involved person in designing and adding the externalities.

About the single syscall-like entrypoint: I actually had some thoughts about this kind of design for srml-contracts wasm environment. This kind of build up would add some inflexibility and overhead into the substrate host interface. This doesn't just harms the compilation efficiency, but also introduces more high-level problems. For instance, if your runtime wants to call a function that takes two buffers and you have them at hand you cannot pass them directly and would have to encode them probably (re)allocating more than once.

Also, there is another special thing about this design: it has properties of dynamic dispatch. The default way of interaction between a wasm instance and the host is by resolving **all** imports at the instantiation time and if the instance imports something that doesn't exist the instance cannot be instantiated. On the other hand, with such a dynamic dispatch it would instantiate successfully (cause it imports only the syscall entrypoint) but when it would fail only at the runtime when it tries to call a non-supported function. In general this can be a feature or an issue, but I am inclined to think that in the case of substrate it is more an issue since I doubt that there is a lot to do with conditional logic (the runtime should behave the same regardless the set of available functions).

But yeah, it would definitely make the interface simpler.
As a reference, there is https://github.com/paritytech/substrate/issues/2334 to achieve the same (but arguably more complex than what you propose)

pepyakin commented on core/sr-io/without_std.rs:
```diff
@@ -385,6 +385,20 @@ pub mod ext {
+		///
+		/// # Returns
+		///
+		/// - The encoded `Result<PublicKey encoded to Vec<u8>, ()>`.
```
Why is the rationaly not using `Option`?

cmichi replied:
I just used the same return type as the other existing functions in the API and would leave it like this for consistency.

pepyakin commented on core/sr-io/without_std.rs:
```diff
@@ -385,6 +385,20 @@ pub mod ext {
+		///
+		/// # Returns
+		///
+		/// - The encoded `Result<offchain::LocalNetworkState, ()>`.
```
I think it is worth noting that the ownership of the returned buffer is transferred to the runtime code and thus the runtime is reponsible for freeing it. Also it would useful to note that this is always a properly allocated pointer (and e.g. cannot be NULL), so the runtime code can always rely on it.

Can we have a description on what is `msg_len`, what does the message mean in this context, is the ownership of this pointer is transferred or just lended, etc.

pepyakin commented on core/sr-io/without_std.rs:
```diff
@@ -385,6 +385,20 @@ pub mod ext {
+		///
+		/// # Returns
+		///
+		/// - The encoded `Result<PublicKey encoded to Vec<u8>, ()>`.
```
- Can we have a description what does `crypto` mean here?
- The same comments as for `ext_local_network_state` regarding `msg_len` and semantics of the return value

pepyakin commented on core/executor/src/wasm_executor.rs:
```diff
@@ -767,6 +768,42 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
+		written_out: *mut u32
+	) -> *mut u8 => {
+		let kind = offchain::CryptoKind::try_from(kind)
+			.map_err(|_| "crypto kind OOB while ext_decrypt: wasm")?;
```
```suggestion
			.map_err(|_| "crypto kind OOB while ext_local_authority_pubkey: wasm")?;
```

pepyakin replied:
Wait, is it acually OOB?

cmichi replied:
Yeah, I checked again. Also I just copied this line from the analog methods.

pepyakin commented on core/executor/src/wasm_executor.rs:
```diff
@@ -767,6 +768,42 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
+
+		let res = this.ext.offchain()
+			.map(|api| api.local_authority_pubkey(kind))
+			.ok_or_else(|| "Calling unavailable API ext_decrypt: wasm")?;
```
```suggestion
			.ok_or_else(|| "Calling unavailable API ext_local_authority_pubkey: wasm")?;
```

pepyakin commented on core/executor/src/wasm_executor.rs:
```diff
@@ -767,6 +768,42 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
+		let len = encoded.len() as u32;
+		let offset = this.heap.allocate(len)? as u32;
+		this.memory.set(offset, &encoded)
+			.map_err(|_| "Invalid attempt to set memory in ext_http_response_headers")?;
```
```suggestion
			.map_err(|_| "Invalid attempt to set memory in ext_local_authority_pubkey")?;
```

pepyakin commented on core/executor/src/wasm_executor.rs:
```diff
@@ -767,6 +768,42 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
+		this.memory.set(offset, &encoded)
+			.map_err(|_| "Invalid attempt to set memory in ext_http_response_headers")?;
+		this.memory.write_primitive(written_out, len)
+			.map_err(|_| "Invalid attempt to write written_out in ext_http_response_headers")?;
```
```suggestion
			.map_err(|_| "Invalid attempt to write written_out in ext_local_authority_pubkey")?;
```

pepyakin commented on core/executor/src/wasm_executor.rs:
```diff
@@ -767,6 +768,42 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
 
 		Ok(offset)
 	},
+	ext_local_network_state(msg_len: *mut u32) -> *mut u8 => {
```
Why `msg_len`?

pepyakin commented on core/executor/src/wasm_executor.rs:
```diff
@@ -767,6 +768,42 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
+		this.memory.set(offset, &encoded)
+			.map_err(|_| "Invalid attempt to set memory in ext_local_network_state")?;
+
+		this.memory.write_primitive(msg_len, len)
```
I think it would be better if we change the name of `msg_len` to something difference, since I find this line quite confusing. `written_out`, as used in the sibiling function, works better IMO.

cmichi commented on core/network/src/service.rs:
```diff
@@ -249,6 +256,12 @@ impl<B: BlockT + 'static, S: NetworkSpecialization<B>, H: ExHashT> NetworkWorker
 	}
 
+	/// Returns the local external addresses.
+	pub fn external_addresses(&self) -> HashSet<Multiaddr> {
```
I've introduced a `NetworkStateInfo` and it seems that it's currently in stable Rust not possible to return `impl Trait` in a trait definition (https://stackoverflow.com/a/39490692). So I have left it out for now.

cmichi commented on core/sr-primitives/src/testing.rs:
```diff
@@ -44,6 +44,13 @@ impl TypedKey for UintAuthorityId {
+impl AsRef<[u8]> for UintAuthorityId {
+	fn as_ref(&self) -> &[u8] {
+		let ptr = self.0 as *const _;
+		unsafe { std::slice::from_raw_parts(ptr, 8) }
```
Yep, `transmute` would work as well and actually I had `unsafe { &std::mem::transmute::<_, &[u8; 8]>(&self.0)[..] }` in a first version, but one other reviewer had complained about transmute. Hence I changed it.

cmichi commented on srml/im-online/src/lib.rs:
```diff
@@ -0,0 +1,305 @@
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
```
I added the missing docs in the meantime, so it's obsolete.

tomusdrw commented on core/offchain/Cargo.toml:
```diff
@@ -16,6 +16,9 @@ parking_lot = "0.8.0"
 runtime_primitives = { package = "sr-primitives", path = "../../core/sr-primitives" }
 transaction_pool = { package = "substrate-transaction-pool", path = "../../core/transaction-pool" }
+network = { package = "substrate-network", path = "../../core/network" }
+serde = { version = "1.0", features = ["derive"] }
```
Is `serde` and `bincode` still needed for anything?

tomusdrw commented on core/offchain/src/api.rs:
```diff
@@ -158,6 +162,33 @@ impl<Storage, KeyProvider> OffchainExt for Api<Storage, KeyProvider> where
 	}
 
+	fn authority_pubkey(&self, kind: CryptoKind) -> Result<Vec<u8>, ()> {
+		match self.read_key(None, kind) {
```
```suggestion
		let key = self.read_key(None, kind)?;
		match key {
		 ...
		}
```

tomusdrw commented on core/offchain/src/api.rs:
```diff
@@ -158,6 +162,33 @@ impl<Storage, KeyProvider> OffchainExt for Api<Storage, KeyProvider> where
+	fn network_state(&self) -> Result<OpaqueNetworkState, ()> {
+		let external_addresses = self.network_state
+			.external_addresses()
+			.into_iter()
```
why do we `into_iter()` and `collect()`? Maybe better return `Vec` from `NetworkStateInfo` or have associated type there?

cmichi replied:
It's a `HashSet` in `core/network`, but I transform to a `Vec`, since `Encode/Decode` is only implemented for `Vec` currently and the support for `Vec` in WASM is really good. I liked your suggestion of returning `impl Iterator<Item = Multiaddr>`, but as written in the reply for that suggestion that's sadly not possible with Rust in traits currently.

Do you see a shorter way of getting around this allocation? The only alternative I see would be to implement `Encode/Decode` for `HashSet` as well and see if/how well it's supported in WASM.

tomusdrw replied:
Well you can always have an associated type and collect to that, but it might be a bit cumbersome to handle.

But maybe we can just change `HashMap` to `Vec`, since it seems we already get an iterator from `Swarm` so you'd just collect into a `Vec` on that level. I guess collecting to a `HashSet` allows you to deduplicate, but is that needed?

cmichi replied:
Yes, good point. I talked to Pierre and the related code just uses a `HashSet` so that it gets serialized into a JavaScript object as opposed to a Map. So we can collect into a `Vec`.

tomusdrw commented on core/offchain/src/api.rs:
```diff
@@ -285,6 +316,72 @@ impl<Storage, KeyProvider> OffchainExt for Api<Storage, KeyProvider> where
+
+impl From<NetworkState> for OpaqueNetworkState {
+	fn from(state: NetworkState) -> OpaqueNetworkState {
+		let enc = Encode::encode(&state.peer_id.as_bytes().to_vec());
```
Are you sure we need both `as_bytes()` and `to_vec()`? This seems like unneeded allocation.

tomusdrw commented on core/offchain/src/api.rs:
```diff
@@ -285,6 +316,72 @@ impl<Storage, KeyProvider> OffchainExt for Api<Storage, KeyProvider> where
+	fn from(state: OpaqueNetworkState) -> NetworkState {
+		let inner_vec = state.peer_id.0;
+
+		let bytes: Vec<u8> = <Vec<u8>>::decode(&mut &inner_vec[..])
```
```suggestion
		let bytes: Vec<u8> = Decode::decode(&mut &inner_vec[..])
```

or 

`let bytes = <<Vec<u8>>::decode(&mut &inner_vec[..])`

tomusdrw commented on core/offchain/src/api.rs:
```diff
@@ -285,6 +316,72 @@ impl<Storage, KeyProvider> OffchainExt for Api<Storage, KeyProvider> where
+		let bytes: Vec<u8> = <Vec<u8>>::decode(&mut &inner_vec[..])
+			.expect("It's always a valid Vec<u8> which is encoded; qed");
+		let peer_id = PeerId::from_bytes(bytes)
+			.expect("These bytes were exported from a valid PeerId; qed");
```
I really think we should never panic here. Let's implement `TryFrom` instead and handle the error.
There are multiple ways how you can end up with `OpaqueNetworkState` that does not represent a valid `NetworkState` at all.

tomusdrw commented on core/sr-io/src/lib.rs:
```diff
@@ -239,6 +240,13 @@ export_api! {
+		fn network_state() -> Result<OpaqueNetworkState, ()>;
+
+		/// Returns the currently configured authority public key, if available.
+		// TODO [ToDr] change into crypto_pubkey(&self, key: Option<CryptoKeyId>, kind: CryptoKind)
```
We should have an issue for this one too.

tomusdrw commented on srml/im-online/src/lib.rs:
```diff
@@ -0,0 +1,442 @@
+				},
+			};
+			if next_gossip < now && not_yet_gossipped {
+				set_worker_status::<T>(now, false);
```
This is subject to race conditions. I think we should use `compare_and_set` from offchain workers API.

cmichi replied:
Created a follow-up (https://github.com/paritytech/substrate/issues/3143) and assigned to myself after private conversation.

tomusdrw commented on srml/im-online/src/lib.rs:
```diff
@@ -0,0 +1,442 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
```
Do we have tests for this module? Or is it coming in a follow up PR?

cmichi replied:
Will be a follow-up PR, I'm already working on it.

gavofyork commented on core/primitives/src/offchain.rs:
```diff
@@ -241,6 +276,12 @@ pub trait Externalities {
+	/// Returns information about the local node's network state.
+	fn network_state(&self) -> Result<OpaqueNetworkState, ()>;
+
+	/// Returns the locally configured authority public key, if available.
```
What is the "authority public key"? There are several keys in use for authorities now.

cmichi replied:
There exists [a trait](https://github.com/paritytech/substrate/blob/master/core/offchain/src/lib.rs#L61-L65) for that with [this implementation](https://github.com/paritytech/substrate/blob/master/core/service/src/lib.rs#L885-L908), which I call to get the current authority public key. This whole part is currently getting reworked in https://github.com/paritytech/substrate/pull/3034, so I thought it's fine in master?

gavofyork replied:
ok

dvc94ch replied:
I started refactoring the offchain workers api (discussed with @tomusdrw) while @tomaka is refactoring the service factory/components. I'm on vacation till next week. The plan is to rebase #3034 on those PRs

gavofyork commented on node/runtime/src/lib.rs:
```diff
@@ -379,6 +379,20 @@ impl sudo::Trait for Runtime {
+	type SessionsPerEra = SessionsPerEra;
+	type UncheckedExtrinsic = UncheckedExtrinsic;
+
+	fn is_valid_authority_id(authority_id: &<Self as im_online::Trait>::AuthorityId) -> bool {
```
this is irregular. instead, create a new `trait IsMember<AuthorityId>`, have `Aura` and `Babe` implement it and have the `im_online::Trait` contain an extra item `type IsValidAuthorityId: IsMember<Self::AuthorityId>;`. here there should just be a `type IsValidAuthorityId = Aura;`.

gavofyork commented on core/offchain/Cargo.toml:
```diff
@@ -1,7 +1,7 @@
 description = "Substrate offchain workers"
 name = "substrate-offchain"
-version = "2.0.0"
+version = "2.1.0"
```
why? version should stay as blanket 2.0.0 until 2.0.0 is ready for release.

cmichi commented on srml/im-online/src/lib.rs:
```diff
@@ -0,0 +1,442 @@
+		fn heartbeat(
+			origin,
+			heartbeat: Heartbeat<T::BlockNumber, T::AuthorityId>,
+			_signature: Vec<u8>
```
`_signature: Vec<u8>,` results in a number of errors, `_signature: Vec<u8>` compiles fine. I suspect that there might be an issue with the macro, are you aware of any issue?

cmichi replied:
Regarding checking: `validate_unsigned(…)` verifies it.

cmichi commented on core/network/src/service.rs:
```diff
@@ -251,6 +256,12 @@ impl<B: BlockT + 'static, S: NetworkSpecialization<B>, H: ExHashT> NetworkWorker
 
+	/// Returns the local external addresses.
+	pub fn external_addresses(&self) -> Vec<Multiaddr> {
+		let swarm = &self.network_service;
```
To provide a stable getter for this property. There is an existing `NetworkState` struct, but it's API is highly unstable and it shouldn't be used.

cmichi replied:
Good spotting, it was a survivor from an earlier version of the PR. I removed it now.

cmichi commented on core/network/src/service.rs:
```diff
@@ -520,6 +562,8 @@ enum ServerToWorkerMsg<B: BlockT, S: NetworkSpecialization<B>> {
 /// You are encouraged to poll this in a separate background thread or task.
 #[must_use = "The NetworkWorker must be polled in order for the network to work"]
 pub struct NetworkWorker<B: BlockT + 'static, S: NetworkSpecialization<B>, H: ExHashT> {
+	/// Updated by the `NetworkWorker` and loaded by the `NetworkService`.
```
Here: https://github.com/paritytech/substrate/pull/3079/files#diff-a045635f3eec17c06b3c7ac7944acceaR500, do you have concerns regarding the name "loading" being associated with the method `load()`?

cmichi commented on core/network/src/service.rs:
```diff
@@ -86,6 +87,8 @@ impl ReportHandle {
 	/// Number of peers we're connected to.
 	num_connected: Arc<AtomicUsize>,
+	/// The local external addresses.
+	external_addresses: Arc<Mutex<Vec<Multiaddr>>>,
```
Thx, fixed.
